odbc/src/api/data.rs

```diff
@@ -0,0 +1,647 @@
+            buffer_len as usize,
+            str_len_or_ind_ptr as *mut _,
+        ),
+        CDataType::Bit => {
+            set_output_fixed_data(&data.to_bool(), target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::Double => {
+            set_output_fixed_data(&data.to_f64(), target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::Float => {
+            set_output_fixed_data(&data.to_f32(), target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::SBigInt | CDataType::Numeric => {
+            set_output_fixed_data(&data.to_i64(), target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::SLong => {
+            set_output_fixed_data(&data.to_i32(), target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::TimeStamp | CDataType::TypeTimestamp => {
+            let dt = data.to_date();
+            let data = Timestamp {
+                year: dt.year() as i16,
+                month: dt.month() as u16,
+                day: dt.day() as u16,
+                hour: dt.hour() as u16,
+                minute: dt.minute() as u16,
+                second: dt.second() as u16,
+                fraction: (dt.nanosecond() as f32 * 0.000001) as u32,
+            };
+            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::Time | CDataType::TypeTime => {
+            let dt = data.to_date();
+            let data = Time {
+                hour: dt.hour() as u16,
+                minute: dt.minute() as u16,
+                second: dt.second() as u16,
+            };
+            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::Date | CDataType::TypeDate => {
+            let dt = data.to_date();
+            let data = Date {
+                year: dt.year() as i16,
+                month: dt.month() as u16,
+                day: dt.day() as u16,
+            };
+            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+        }
+        _ => {
+            mongo_handle.add_diag_info(ODBCError::Unimplemented("unimplemented data type"));
+            SqlReturn::ERROR
+        }
+    }
+}
+
+///
+/// input_text_to_string converts an input cstring to a rust String.
+/// It assumes nul termination if the supplied length is negative.
+///
+/// # Safety
+/// This converts raw C-pointers to rust Strings, which requires unsafe operations
+///
+/// for now this is cfg(test), when/if we decide to provide an ascii driver this will
+/// see use in actual code.
+#[allow(clippy::uninit_vec)]
+#[cfg(test)]
+pub unsafe fn input_text_to_string(text: *const Char, len: usize) -> String {
+    if (len as isize) < 0 {
+        let mut dst = Vec::new();
+        let mut itr = text;
+        {
+            while *itr != 0 {
+                dst.push(*itr);
+                itr = itr.offset(1);
+            }
+        }
+        return String::from_utf8_unchecked(dst);
```

Comment:
This doesn't check that the passed string is valid utf-8. Since this is intended to be used for ASCII text _only_, this has to be valid utf-8.